### PR TITLE
Implement a behaviour driven fetch api.

### PR DESCRIPTION
**Pull request by oferrigni**
Allow individual object records to own their fetch implementation.

### CODE REVIEW

File: src/chef_sql.erl [hunk 1]
```diff
@@ -680,15 +680,32 @@ statements() ->
...
-fetch_sandbox(OrgId, SandboxID) ->
-    case sqerl:select(find_sandbox_by_id, [OrgId, SandboxID]) of
-        {ok, Rows} when is_list(Rows) ->
-            {ok, sandbox_join_rows_to_record(Rows)};
-        {ok, none} ->
-            {ok, not_found};
-        {error, Error} ->
-            {error, Error}
-    end.
+fetch(Record) ->
```
Comment (seth) [inline src/chef_sql.erl:682]: Let's add a spec to `fetch/1`.

File: src/chef_sql.erl [hunk 2]
```diff
@@ -680,15 +680,32 @@ statements() ->
...
+        {Query :: atom(), BindParameters :: list() } |
+        {Query :: atom(), BindParameters :: list(),
+                             Fields :: list(atom())}) ->
+                                    {ok, none} |
+                                    {ok, list()} |
+                                    {error, term()}.
+
+select_rows({Query, BindParameters}) ->
+    match_result(sqerl:select(Query, BindParameters));
+select_rows({Query, BindParameters, Transform}) when is_tuple(Transform) ->
```
Comment (seth) [inline src/chef_sql.erl:697]: Are you sure `Transform` is always going to be a tuple?

Comment (oferrigni) [inline src/chef_sql.erl:697]: The tuple includes a lead atom that describes the transform.  I didn't see any other usages that used a different form.

File: src/chef_sql.erl [hunk 3]
```diff
@@ -678,17 +678,39 @@ statements() ->
     {ok, Statements} = file:consult(Path),
     Statements.
 
-%% Sandbox Operations
+-spec(fetch(chef_object:object_rec()) -> chef_object:object_rec()).
+fetch(Record) ->
+    chef_object:fetch(Record, fun select_rows/1).
+
+-spec select_rows(
+
```
Comment (seth) [inline src/chef_sql.erl]: format niggle: remove this empty line

File: src/chef_sql.erl [hunk 4]
```diff
@@ -678,17 +678,39 @@ statements() ->
     {ok, Statements} = file:consult(Path),
     Statements.
 
-%% Sandbox Operations
+-spec(fetch(chef_object:object_rec()) -> chef_object:object_rec()).
+fetch(Record) ->
+    chef_object:fetch(Record, fun select_rows/1).
+
+-spec select_rows(
```
Comment (seth) [inline src/chef_sql.erl:685]: I like how this turned out. Much easier to read in this style.

File: src/chef_sql.erl [hunk 5]
```diff
@@ -678,17 +678,39 @@ statements() ->
...
+      ReturnFieldNames :: [atom()],
+      ReturnTransform :: tuple().
+
+select_rows({Query, BindParameters}) ->
+    match_result(sqerl:select(Query, BindParameters));
+select_rows({Query, BindParameters, Transform}) when is_tuple(Transform) ->
+    match_result(sqerl:select(Query, BindParameters, Transform));
+select_rows({Query, BindParameters, Fields = [_|_]}) ->
+    match_result(sqerl:select(Query, BindParameters, rows_as_scalars, Fields)).
+-spec match_result(Input) -> NormalizedResult when
```
Comment (seth) [inline src/chef_sql.erl:702]: Format: add newline to separate end of select_rows function and start of match_result spec.

**Comment by seth** [comment]
a few small formatting suggestions, but overall :+1:

**Pull request merged**